python/pydantic_core/_pydantic_core.pyi

```diff
@@ -822,54 +822,280 @@ class ValidationError(ValueError):
 
 @final
 class PydanticCustomError(ValueError):
+    """A custom exception providing flexible error handling for Pydantic validators.
+
+    You can raise this error in custom validators when you'd like flexibility in regards to the error type, message, and context.
+
+    Arguments:
+        error_type: The error type, typically a snake_case string.
+        message_template: The message template, formatted with any context variables in `{curly_braces}`.
+        context: The data to inject into the message template.
+
+    Example:
+        ```py
+        from pydantic_core import PydanticCustomError
+
+        def custom_validator(v) -> None:
+            if v <= 10:
+                raise PydanticCustomError('custom_value_error', 'Value must be greater than {value}', {'value': 10, 'extra_context': 'extra_data'})
+            return v
+        ```
+    """
+
     def __new__(
         cls, error_type: LiteralString, message_template: LiteralString, context: dict[str, Any] | None = None
-    ) -> Self: ...
+    ) -> Self:
+        """Create a new `PydanticCustomError`."""
+
     @property
-    def context(self) -> dict[str, Any] | None: ...
+    def context(self) -> dict[str, Any] | None:
+        """Values which are required to render the error message, and could hence be useful in passing error data forward."""
+
     @property
-    def type(self) -> str: ...
+    def type(self) -> str:
+        """The error type associated with the error. For consistency with Pydantic, this is typically a snake_case string."""
+
     @property
-    def message_template(self) -> str: ...
-    def message(self) -> str: ...
+    def message_template(self) -> str:
+        """The message template associated with the error. This is a string that can be formatted with context variables in `{curly_braces}`."""
+
+    def message(self) -> str:
+        """The formatted message associated with the error. This presents as the message template with context variables appropriately injected."""
 
 @final
 class PydanticKnownError(ValueError):
-    def __new__(cls, error_type: ErrorType, context: dict[str, Any] | None = None) -> Self: ...
+    """A helper class for raising exceptions that mimic Pydantic's built-in exceptions, with more flexibility in regards to context.
+
+    Unlike [`PydanticCustomError`][pydantic_core.PydanticCustomError], the `error_type` argument must be a known [`ErrorType`][pydantic_core.ErrorType].
+
+    Arguments:
+        error_type: The error type, one of the known `ErrorType`s from `pydantic_core`.
```

Comment:
```suggestion
        error_type: The error type.
```

Or with a proper reference to `ErrorType`. But as I said, it doesn't really add value to have it repeated (API docs will already show the link, and IDEs have the type hint).